zkevm-circuits/src/state_circuit.rs

```diff
@@ -41,14 +41,21 @@ const N_LIMBS_ID: usize = 2;
 /// Config for StateCircuit
 #[derive(Clone)]
 pub struct StateCircuitConfig<F> {
-    selector: Column<Fixed>, // Figure out why you get errors when this is Selector.
+    // Figure out why you get errors when this is Selector.
+    selector: Column<Fixed>,
     // https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/407
     rw_table: RwTable,
     mpt_table: MptTable,
     sort_keys: SortKeysConfig,
-    initial_value: Column<Advice>, /* Assigned value at the start of the block. For Rw::Account
-                                    * and Rw::AccountStorage rows this is the committed value in
-                                    * the MPT, for others, it is 0. */
+    // Assigned value at the start of the block. For Rw::Account and
+    // Rw::AccountStorage rows this is the committed value in the MPT, for
+    // others, it is 0.
+    initial_value: Column<Advice>,
+    // For Rw::AccountStorage, identify if committed value or new value is zero.
+    // Will do lookup for ProofType::StorageDoesNotExist if both are zero,
+    // otherwise do lookup for ProofType::StorageChanged (either is non-zero).
+    is_initial_value_zero: Column<Advice>,
```

Comment:
Can you combine these two with a `BatchedIsZeroGadget`?

Reply:
It would be better to use the `BatchedIsZeroGadget` in zkevm-circuits/src/evm_circuit/util/math_gadget.rs.

Reply:
I tried to replace this field with `BatchedIsZeroGadget`, but it seems that the first parameter of function [BatchedIsZeroGadget::construct](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/evm_circuit/util/math_gadget.rs#L102) is `evm_circuit::constraint_builder::ConstraintBuilder` (not `state_circuit::constraint_builder::ConstraintBuilder`).

And it seems that the `state_circuit::constraint_builder::ConstraintBuilder` needs this `StateCircuitConfig` (this field belongs to) to [construct here](https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/state_circuit.rs#L135).

@z2trillion  Any suggestion to refactor `BatchedIsZeroGadget` or `state_circuit::constraint_builder::ConstraintBuilder`? Thanks.

Reply:
Sorry about this. I didn't realize that `BatchedIsZeroGadget` requires the evm `ConstraintBuilder` to construct. Can you just add a todo here: "TODO: use BatchedIsZeroGadget here once it no longer requires the evm circuit constraint builder."

Reply:
Yes. I see. Thanks 🙏.
I will add `TODO: use BatchedIsZeroGadget here once it no longer requires the evm circuit constraint builder.` in the `BALANCE` circuit PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/683 (which will use this non-existing proof code).

Reply:
I saw the PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/961. Thanks.